File: backend/lib/backend/game/state.ex
defmodule Backend.Game.State do
  @moduledoc """
  Shape completo do estado do jogo (porta de frontend/src/engine/state/GameState.ts).

  Single source of truth. Tudo que pode ser salvo, carregado, renderizado ou
  sincronizado passa por aqui. Todos os modules de game logic recebem e retornam
  esta struct.
  """

  @save_version 2

  @type era_id :: 1..8

  @type resources :: %{
          tokens: float(),
          compute: float(),
          data: float(),
          funding: float(),
          hype: float(),
          hallucinations: float()
        }

  @type t :: %__MODULE__{
          version: integer(),
          player_id: String.t() | nil,
          display_name: String.t(),
          created_at: integer(),
          last_tick: integer(),
          tick_count: integer(),
          resources: resources(),
          total_tokens_earned: float(),
          buildings: %{atom() => integer()},
          upgrades: [String.t()],
          era: era_id(),
          prestige_count: integer(),
          insight_points: integer(),
          permanent_upgrades: [String.t()],
          reborn_count: integer(),
          reborn_points: integer(),
          unlocked_perks: [String.t()],
          total_prestiges_all_time: integer(),
          ml_steps_trained: integer(),
          ml_capability_score: integer(),
          achievements: [String.t()]
        }

  defstruct version: @save_version,
            player_id: nil,
            display_name: "Anonymous Founder",
            created_at: 0,
            last_tick: 0,
            tick_count: 0,
            resources: %{
              tokens: 0.0,
              compute: 0.0,
              data: 0.0,
              funding: 0.0,
              hype: 0.0,
              hallucinations: 0.0
            },
            total_tokens_earned: 0.0,
            buildings: %{},
            upgrades: [],
            era: 1,
            prestige_count: 0,
            insight_points: 0,
            permanent_upgrades: [],
            reborn_count: 0,
            reborn_points: 0,
            unlocked_perks: [],
            total_prestiges_all_time: 0,
            ml_steps_trained: 0,
            ml_capability_score: 0,
            achievements: []

  @doc "Versão atual do save format."
  def save_version, do: @save_version

  @doc "Cria um estado fresh (reset/prestige/reborn)."
  @spec new(String.t() | nil) :: t()
  def new(player_id \\ nil) do
    now = System.system_time(:millisecond)

    %__MODULE__{
      player_id: player_id,
      created_at: now,
      last_tick: now
    }
  end

  @doc "Serializa pra map JSON-friendly (pra Postgres jsonb ou Phoenix push)."
  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = state) do
    state
    |> Map.from_struct()
    |> Map.update!(:buildings, fn buildings ->
      Map.new(buildings, fn {k, v} -> {Atom.to_string(k), v} end)
    end)
  end

  @doc "Reidratatação from map (jsonb do Postgres → struct)."
  @spec from_map(map()) :: t()
  def from_map(map) when is_map(map) do
    base = %__MODULE__{}

    attrs =
      map
      |> Enum.map(fn {k, v} ->
        key = if is_atom(k), do: k, else: String.to_existing_atom(to_string(k))
        {key, v}
      end)
      |> Map.new()

    buildings =
      case Map.get(attrs, :buildings, %{}) do
        b when is_map(b) ->
          Map.new(b, fn {k, v} ->
            key = if is_atom(k), do: k, else: String.to_atom(to_string(k))
            {key, v}
          end)

        _ ->
          %{}
      end

    resources =
      case Map.get(attrs, :resources, base.resources) do
        r when is_map(r) ->
          Map.new(base.resources, fn {key, default} ->
            string_key = Atom.to_string(key)
            value = Map.get(r, key) || Map.get(r, string_key) || default
            {key, to_float(value)}
          end)

        _ ->
          base.resources
      end

    struct(base, Map.merge(attrs, %{buildings: buildings, resources: resources}))
  end

  defp to_float(v) when is_integer(v), do: v * 1.0
  defp to_float(v) when is_float(v), do: v
  defp to_float(_), do: 0.0

  @doc "Quanto de um building específico o player tem."
  @spec owned(t(), atom()) :: integer()
  def owned(%__MODULE__{buildings: buildings}, building_id) when is_atom(building_id) do
    Map.get(buildings, building_id, 0)
  end

  @doc "Aplica um update imutável no struct (conveniência para pipelines)."
  @spec update(t(), (t() -> t())) :: t()
  def update(%__MODULE__{} = state, fun) when is_function(fun, 1), do: fun.(state)
end

File: server/src/game/state.rs
//! GameState — shape canônico do estado do jogo.
//!
//! Porta 1:1 da interface `GameState` do concept doc + frontend v1. Qualquer
//! mudança aqui deve ser refletida no `GAME_CONCEPT.md` e nas fórmulas.
//!
//! O struct é `Serialize + Deserialize + TS` para:
//! - Persistir em Postgres como JSONB (`sqlx::types::Json`)
//! - Enviar pro client via WebSocket como JSON
//! - Gerar `GameState.ts` automaticamente com `ts-rs`

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use ts_rs::TS;

use super::data::buildings::BuildingId;
use super::data::eras::EraId;

pub const SAVE_VERSION: u32 = 2;

/// Seis recursos interligados. Ver `GAME_CONCEPT.md` para a dinâmica.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../client/src/bindings/")]
pub struct Resources {
    pub tokens: f64,
    pub compute: f64,
    pub data: f64,
    pub funding: f64,
    pub hype: f64,
    /// 0.0–1.0. Em 1.0, drena 100% da produção e dispara "rogue" event.
    pub hallucinations: f64,
}

impl Default for Resources {
    fn default() -> Self {
        Self {
            tokens: 0.0,
            compute: 0.0,
            data: 0.0,
            funding: 0.0,
            hype: 0.0,
            hallucinations: 0.0,
        }
    }
}

/// Estado completo de um jogador. Única source of truth.
///
/// **Invariante**: todos os campos devem existir em `GAME_CONCEPT.md` e
/// ter a mesma semântica. Campos novos requerem bump de `SAVE_VERSION`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../client/src/bindings/")]
pub struct GameState {
    pub version: u32,
    pub player_id: String,
    pub display_name: String,

    /// Timestamps em ms epoch.
    pub created_at: i64,
    pub last_tick: i64,
    pub tick_count: u64,

    pub resources: Resources,
    pub total_tokens_earned: f64,

    /// Mapa building → quantidade. String key pra JSON-friendly.
    pub buildings: HashMap<String, u32>,
    /// IDs de upgrades comprados (one-shot).
    pub upgrades: Vec<String>,

    pub era: EraId,

    pub prestige_count: u32,
    pub insight_points: u32,
    /// Array com stacks (IDs podem repetir).
    pub permanent_upgrades: Vec<String>,

    pub reborn_count: u32,
    pub reborn_points: u32,
    pub unlocked_perks: Vec<String>,
    pub total_prestiges_all_time: u32,

    /// Sincronizado com o Python `/evaluate`.
    pub ml_steps_trained: u32,
    pub ml_capability_score: u64,

    pub achievements: Vec<String>,
}

impl GameState {
    /// Cria um novo estado fresh. Usado em first-time join, prestige e reborn.
    pub fn new(player_id: String) -> Self {
        let now = now_ms();
        Self {
            version: SAVE_VERSION,
            player_id,
            display_name: "Anonymous Founder".to_string(),
            created_at: now,
            last_tick: now,
            tick_count: 0,
            resources: Resources::default(),
            total_tokens_earned: 0.0,
            buildings: HashMap::new(),
            upgrades: Vec::new(),
            era: EraId::Hardcoded,
            prestige_count: 0,
            insight_points: 0,
            permanent_upgrades: Vec::new(),
            reborn_count: 0,
            reborn_points: 0,
            unlocked_perks: Vec::new(),
            total_prestiges_all_time: 0,
            ml_steps_trained: 0,
            ml_capability_score: 0,
            achievements: Vec::new(),
        }
    }

    /// Quanto de um building o player tem.
    pub fn owned(&self, id: BuildingId) -> u32 {
        self.buildings.get(id.as_str()).copied().unwrap_or(0)
    }

    /// Incrementa a quantidade de um building.
    pub fn increment_building(&mut self, id: BuildingId, by: u32) {
        let entry = self.buildings.entry(id.as_str().to_string()).or_insert(0);
        *entry += by;
    }

    /// Adiciona tokens ao state. Mantém o total_tokens_earned em sync.
    pub fn add_tokens(&mut self, amount: f64) {
        self.resources.tokens += amount;
        self.total_tokens_earned += amount;
    }
}

/// Helper: timestamp ms atual.
pub fn now_ms() -> i64 {
    chrono::Utc::now().timestamp_millis()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new_state_has_zero_resources() {
        let s = GameState::new("test".to_string());
        assert_eq!(s.resources.tokens, 0.0);
        assert_eq!(s.era, EraId::Hardcoded);
        assert_eq!(s.version, SAVE_VERSION);
    }

    #[test]
    fn add_tokens_updates_total_earned() {
        let mut s = GameState::new("test".to_string());
        s.add_tokens(100.0);
        assert_eq!(s.resources.tokens, 100.0);
        assert_eq!(s.total_tokens_earned, 100.0);
    }

    #[test]
    fn increment_building_defaults_to_zero() {
        let mut s = GameState::new("test".to_string());
        assert_eq!(s.owned(BuildingId::IfElseBot), 0);
        s.increment_building(BuildingId::IfElseBot, 3);
        assert_eq!(s.owned(BuildingId::IfElseBot), 3);
        s.increment_building(BuildingId::IfElseBot, 2);
        assert_eq!(s.owned(BuildingId::IfElseBot), 5);
    }
}
